### PR TITLE
feat(use-throttle): throttle a given function

### PR DESCRIPTION
**Pull request by yossi-eynav**

### CODE REVIEW

**Comment by getify** [comment]
This is a good start, thanks! I think there's a few things we should address.

First, you're re-creating a `throttleFunction(..)` each time `useThrottle(..)` gets called, which is a bit inefficient. Ideally, I think that function can be created once and stored in the slot alongside the original function, timer, and last-execution timestamp. On subsequent calls, it would just be returned.

Second, I think we can simplify by just using `useState(..)` instead of managing a state slot manually... something like this:

```js
function useThrottle(fn,timer) {
   var slot = useState({ fn, timer, lastExecuted });
   if (!slot.throttledFn) {
      slot.throttledFn = function throttledFn(...args) {
         // timestamp check
         // ..
         // execute original function, pass along any args
         return slot.fn.apply(this,args);
      };
   }
   return slot.throttledFn;
}
```

Third, this part needs to be handled:

> These two hooks also need to support an optional third parameter, a guards-list, similar to `useEffect(..)` and `useMemo(..)`, which conditionally redefines the debounced/throttled function if the guards have changed (aka, if closure is required).

Could be done sorta like this:

```js
function useThrottle(fn,timer,...guards) {
   // passed in any guards?
   if (guards.length > 0) {
      // only passed a single guards list?
      if (guards.length == 1 && Array.isArray(guards[0])) {
         guards = guards[0];
      }
   }

   var slot = useState({ fn, timer, lastExecuted, });
   if (!slot.throttledFn || guardsChanged(slot.guards,guards)) {
      slot.guards = guards;
      slot.throttledFn = ..
   }
   return slot.throttledFn;
}
```

**Comment by yossi-eynav** [comment]
@getify thanks! check again please, added the guards + refactored it to useState

File: src/tng-hooks.src.js [hunk 1]
```diff
@@ -120,35 +120,38 @@
...
+				if (guards.length === 1 && Array.isArray(guards[0])) {
+					guards = guards[0];
+				}
 			}
 
-			return function throttleFunction(...args) {
-				const [fn, timer, lastExecution] = bucket.stateSlots[slotIdx];
-				const currentTime = Date.now();
+			if (!slot.throttledFn || guardsChanged(slot.guards, guards)) {
+				slot.guards = guards;
```
Comment (getify) [inline src/tng-hooks.src.js]: insert here:

```js
slot.fn = fn;
```

File: src/tng-hooks.src.js [hunk 2]
```diff
@@ -120,35 +120,38 @@
...
+					const currentTime = Date.now();
 
-				if(lastExecution + timer < currentTime) {
-					try {
-						fn(...args);
-					} finally {
-						bucket.stateSlots[slotIdx][2] = currentTime;
+					if(lastExecute + timer < currentTime) {
+						try {
+							fn(...args);
```
Comment (getify) [inline src/tng-hooks.src.js]: Change to `slot.fn.apply(this,args);` so that it uses the updated function (if updated because of the guards changing).

**Comment by yossi-eynav** [comment]
@getify do you want me to close it? it's stale for couple of months

**Comment by getify** [comment]
still planning to come back to it. just been busy on other stuff.

**Pull request closed without merging**